frontend/src/lib/StaticGraphParser.ts

```diff
@@ -76,72 +76,89 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
-        // for incoming edges to only be drawn from its dependencies, not the DAG node itself
-        if (parent && !task.dependencies) {
-          graph.setEdge(parent, nodeId);
-        }
+        buildDag(graph, task.template, templates, alreadyVisited, '');
+        return;
+      }
 
-        // This object contains information about the node that we display to the user when they
-        // click on a node in the graph
-        const info = new SelectedNodeInfo();
-        if (task.when) {
-          info.condition = task.when;
-        }
+      // If this task has already been visited, retrieve the qualified path name that was assigned
+      // to it, add an edge, and move on to the next task
+      if (alreadyVisited.has(task.name)) {
+        graph.setEdge(parentFullPath, alreadyVisited.get(task.name)!);
+        return;
+      }
 
-        // "Child" here is the template that this task points to. This template should either be a
-        // DAG, in which case we recurse, or a container which can be thought of as a leaf node
-        const child = templates.get(task.template);
-        if (child) {
-          if (child.nodeType === 'dag') {
-            buildDag(graph, task.template, templates, nodeId);
-          } else if (child.nodeType === 'container' ) {
-            _populateInfoFromTemplate(info, child.template);
-          } else {
-            logger.error(`Unknown nodetype: ${child.nodeType} on template: ${child.template}`);
-          }
-        }
+      // Parent here will be the task that pointed to this DAG template.
+      // We do not set an edge if the task has dependencies because in that case it makes sense
+      // for incoming edges to only be drawn from its dependencies, not the DAG node itself
+      if (parentFullPath && !task.dependencies) {
+        graph.setEdge(parentFullPath, nodeId);
+      }
 
-        graph.setNode(nodeId, {
-          bgColor: task.when ? 'cornsilk' : undefined,
-          height: Constants.NODE_HEIGHT,
-          info,
-          label: task.template,
-          width: Constants.NODE_WIDTH,
-        });
+      // This object contains information about the node that we display to the user when they
+      // click on a node in the graph
+      const info = new SelectedNodeInfo();
+      if (task.when) {
+        info.condition = task.when;
       }
 
+      // "Child" here is the template that this task points to. This template should either be a
+      // DAG, in which case we recurse, or a container which can be thought of as a leaf node
+      const child = templates.get(task.template);
+      if (child) {
+        if (child.nodeType === 'dag') {
+          buildDag(graph, task.template, templates, alreadyVisited, nodeId);
+        } else if (child.nodeType === 'container' ) {
+          _populateInfoFromTemplate(info, child.template);
+        } else {
+          logger.error(`Unknown nodetype: ${child.nodeType} on template: ${child.template}`);
```

Comment:
Should this error be shown? This will have been a user error (uploading a pipeline with incorrect format), and we should probably surface that.

Reply:
Hm, I think you're right.